File: csharp/bowling/Bowling.cs
public class BowlingGame
{   
    private int[,] _rolls = new int[12, 2];
	private int _frame = 0;
	private int _frameRoll = 0;
    private int _additionalRolls = 0;
    public void Roll(int pins) 
    {
        if(pins < 0 || pins > 10) throw new ArgumentException();
        if(_frame > 9)
        {
            if(IsSpare(9))
            {
                _additionalRolls++;
                if(_additionalRolls > 1) throw new ArgumentException();
            }
            else if(IsStrike(9))
            {
                _additionalRolls++;
                if(_additionalRolls > 2) throw new ArgumentException();
            }
            else throw new ArgumentException();
        }
		_rolls[_frame, _frameRoll] = pins;
		if(_frameRoll == 1 || pins == 10)
		{
            if(_rolls[_frame, 0] + _rolls[_frame, 1] > 10) throw new ArgumentException();
			_frame++;
			_frameRoll = 0;
		}
		else _frameRoll = 1;
    }
    public int? Score()
    {
        if(_frame < 10) throw new ArgumentException();
        if(IsSpare(9) && _additionalRolls != 1) throw new ArgumentException();
        if(IsStrike(9) && _additionalRolls != 2) throw new ArgumentException();
		var result = 0;
		for(var i = 0; i < 10; i++)
		{
			if(IsSpare(i)) result += _rolls[i + 1, 0];
			if(IsStrike(i)) result += StrikeAddition(i + 1);
			
			for(var j = 0; j < _rolls.GetLength(1); j++)
			{
				result += _rolls[i, j];
			}
		}
		return result;
    }
    private bool IsSpare(int frame)
    => _rolls[frame, 0] != 10 && _rolls[frame, 0] + _rolls[frame, 1] == 10;
    private bool IsStrike(int frame)
    => _rolls[frame, 0] == 10;
    private int StrikeAddition(int frame)
    {
        var result = _rolls[frame, 0];
        if(result == 10) result += _rolls[frame + 1, 0];
        else result += _rolls[frame, 1];
        return result;
    }
}

File: javascript/bowling/bowling.js
//
// This is only a SKELETON file for the 'Bowling' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Bowling {
  constructor() {
    this._rolls = [];
  }

  roll(pins) {
    if (pins < 0) {
      throw new Error('Negative roll is invalid');
    }
    if (pins > 10) {
      throw new Error('Pin count exceeds pins on the lane');
    }
    if (this.isGameOver()) {
      throw new Error('Cannot roll after game is over');
    }

    this._rolls.push(pins);
    this.validateCurrentRolls();
  }

  score() {
    if (!this.isGameOver()) {
      throw new Error('Score cannot be taken until the end of the game');
    }

    let totalScore = 0;
    let rollIndex = 0;

    for (let frame = 0; frame < 10; frame++) {
      if (this.isStrike(rollIndex)) {
        totalScore += 10 + this._rolls[rollIndex + 1] + this._rolls[rollIndex + 2];
        rollIndex += 1;
      } else if (this.isSpare(rollIndex)) {
        totalScore += 10 + this._rolls[rollIndex + 2];
        rollIndex += 2;
      } else {
        totalScore += this._rolls[rollIndex] + this._rolls[rollIndex + 1];
        rollIndex += 2;
      }
    }

    return totalScore;
  }

  isStrike(rollIndex) {
    return this._rolls[rollIndex] === 10;
  }
  isSpare(rollIndex) {
    return this._rolls[rollIndex] + this._rolls[rollIndex + 1] === 10;
  }
  
  isGameOver() {
    let rollIndex = 0;

    for (let frame = 0; frame < 10; frame++) {
      if (rollIndex >= this._rolls.length) return false;

      if (frame === 9) {
        if (this.isStrike(rollIndex)) {
          return this._rolls.length === rollIndex + 3;
        }
        if (this._rolls[rollIndex] + this._rolls[rollIndex + 1] === 10) {
          return this._rolls.length === rollIndex + 3;
        }
        return this._rolls.length === rollIndex + 2;
      }

      if (this.isStrike(rollIndex)) {
        rollIndex += 1;
      } else {
        rollIndex += 2;
      }
    }
    return true;
  }

  validateCurrentRolls() {
    let rollIndex = 0;

    for (let frame = 0; frame < 10; frame++) {
      if (rollIndex >= this._rolls.length) break;

      if (frame === 9) {
        const r1 = this._rolls[rollIndex];
        const r2 = this._rolls[rollIndex + 1];
        const r3 = this._rolls[rollIndex + 2];

        if (r1 !== undefined && r2 !== undefined) {
          if (r1 < 10 && r1 + r2 > 10) {
            throw new Error('Pin count exceeds pins on the lane');
          }
        }
        if (r1 === 10 && r2 !== undefined && r3 !== undefined) {
          if (r2 < 10 && r2 + r3 > 10) {
            throw new Error('Pin count exceeds pins on the lane');
          }
        }
        break;
      }

      if (this.isStrike(rollIndex)) {
        rollIndex += 1;
      } else {
        const first = this._rolls[rollIndex];
        const second = this._rolls[rollIndex + 1];
        if (second !== undefined && first + second > 10) {
          throw new Error('Pin count exceeds pins on the lane');
        }
        rollIndex += 2;
      }
    }
  }
}
